Shared/mods/deathmatch/logic/luadefs/CLuaFileDefs.cpp

```diff
@@ -702,51 +702,91 @@ int CLuaFileDefs::fileFlush(lua_State* luaVM)
-    CScriptFile*  pFile;
-    unsigned long ulCount = 0;
+std::variant<bool, std::string> CLuaFileDefs::fileRead (
+    lua_State* luaVM,
+    std::variant<CScriptFile*, std::string> file,
+    std::optional<std::uint32_t> count
+) {
+    const auto& ReadFile = [&](CScriptFile* pFile, std::uint32_t count)
+        -> std::variant<bool, std::string>
+    {
+        if (count == 0)
+            return "";
 
-    CScriptArgReader argStream(luaVM);
-    argStream.ReadUserData(pFile);
-    argStream.ReadNumber(ulCount);
+        SString buffer;
+        auto    bytesRead = pFile->Read(count, buffer);
 
-    if (!argStream.HasErrors())
+        if (bytesRead == -2)
+        {
+            m_pScriptDebugging->LogWarning(luaVM, "out of memory");
+            return false;
+        }
+        if (bytesRead >= 0)
+            return buffer;
+    };
+    if (std::holds_alternative<CScriptFile*>(file))
     {
-        // Reading zero bytes from a file results in an empty string
-        if (ulCount == 0)
+        auto pFile = std::get<CScriptFile*>(file);
+        if (!pFile)
         {
-            lua_pushstring(luaVM, "");
-            return 1;
+            m_pScriptDebugging->LogBadPointer(luaVM, "file", 1);
+            return false;
         }
-
-        // Allocate a buffer to read the stuff into and read some :~ into it
-        SString buffer;
-        long    lBytesRead = pFile->Read(ulCount, buffer);
-
-        if (lBytesRead >= 0)
+        if (!count.has_value())
         {
-            // Push the string onto the Lua stack. Use pushlstring so we are binary
-            // compatible. Normal push string takes zero terminated strings.
-            lua_pushlstring(luaVM, buffer.data(), lBytesRead);
-            return 1;
+            m_pScriptDebugging->LogBadType(luaVM);
+            return false;
         }
-        else if (lBytesRead == -2)
+        return ReadFile(pFile, count.value());
+    }
+
+    CLuaMain* pLuaMain = m_pLuaManager->GetVirtualMachine(luaVM);
+    if (!pLuaMain)
+        return false;
+
+    std::string strInputPath = std::get<std::string>(file);
+
+    SString    strAbsPath;
+    SString    strMetaPath;
+    CResource* pThisResource = pLuaMain->GetResource();
+    CResource* pResource = pThisResource;
+    if (!CResourceManager::ParseResourcePathInput(strInputPath, pResource, &strAbsPath, &strMetaPath))
+        return false;
+
+    {
+        auto canModify = CheckCanModifyOtherResource(pThisResource, pResource);
+        if (!canModify.first)
         {
-            m_pScriptDebugging->LogWarning(luaVM, "out of memory");
+            throw std::runtime_error(canModify.second);
         }
-        else
+    }
+    {
+        auto canModify = CheckCanAccessOtherResourceFile(pThisResource, pResource, strAbsPath);
+        if (!canModify.first)
         {
-            m_pScriptDebugging->LogBadPointer(luaVM, "file", 1);
+            throw std::runtime_error(canModify.second);
```

Comment:
This will lead to std::terminate because runtime_error isn't properly handled by the argument parser.